File: 2º/PDOO/Practias/Napakalaki-ruby/pkg/Napakalaki-ruby-0.0.1/lib/player.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
require_relative 'dice'
require_relative 'card_dealer'
require_relative 'combat_result'

class Player
  attr_accessor :name
  attr_accessor :level
  attr_accessor  :pendingBadConsequence
  attr_accessor :visibleTreasures
  attr_accessor :hiddenTreasures
  attr_accessor :dead
  attr_accessor :enemy
  attr_accessor :canISteal
  
   @@MAXLEVEL = 10
  def initialize(name)
    @name=name
    @level = 0
    @dead = true
    @canISteal =true
    @visibleTreasures = Array.new
    @hiddenTreasures = Array.new
    @enemy = nil
    @pendingBadConsequence = nil
  end
  
  def bringToLife
    @dead= false
  end
  def to_s
    "Name: #{@name}\tLevel: #{@level}"
  end
  def getCombatLevel
    lv=@level
    for j in @visibleTreasures
      lv+=j.bonus
    end
    return lv
  end
  
  def incrementLevels(i)
		@level=@level + i
  end
  
  def decrementLevels(i)
    if (@level > i)
      @level=@level - i
    else
      @level = 0
    end
  end
 
  def setPendingBadConsequence(b)
     @pendingBadConsequence = b
  end
 
  def applyPrize(m)
    nlevels=m.getLevelsGained
    incrementLevels(nlevels)
    ntreasures=m.getTreasuresGained

    if(ntreasures>0)
      dealer=CardDealer.instance
      ntreasures.times do
        treasure=dealer.nextTreasure
        @hiddenTreasures << treasure
      end
    end
  end
  
  def applyBadConsequence(m)
    bad_consequence=m.bc
    n_levels=bad_consequence.levels
    decrementLevels(n_levels)
    pending_bad=bad_consequence.adjustToFitTreasureList(@visibleTreasures,@hiddenTreasures) #4
    setPendingBadConsequence(pending_bad)
  end
  
  def canMakeTreasureVisible(t)
    ret = true
    type = t.type
    num_both = howManyVisibleTreasures(TreasureKind::BOTHHANDS)
    num_one = howManyVisibleTreasures(TreasureKind::ONEHAND)
    if (type == TreasureKind::ONEHAND)
        if ((num_both > 0) || (num_one > 1))
          ret = false
        end
    elsif (type == TreasureKind::BOTHHANDS)
      if ((num_both > 0) || (num_one > 0))
        ret = false
      end
    elsif (howManyVisibleTreasures(type) > 0)
      ret = false
    end
    ret
  end
  
  def howManyVisibleTreasures(tKind)
    	k = 0
		for i in 0..@visibleTreasures.size
			if (@visibleTreasures[i] == tKind)
        k = k+ 1
			end
    end
    k
  end
  
  def dieIfNoTreasures
    
    if( @hiddenTreasures.empty? && @visibleTreasures.empty? )
      @dead = true
    end
  end
  
  def isDead
     return @dead
  end
   
  def combat(m)
    mylevel=getCombatLevel 
    monsterlevel=m.combatLevel

    if(!canISteal)
      dice=Dice.instance
      number=dice.nextNumber
      
      if(number<3)
        e_level=@enemy.getCombatLevel
        monsterlevel+=e_level
      end

    end

    if(mylevel > monsterlevel)
      applyPrize(m)
      if(@level>=@@MAXLEVEL)
        combat_result=CombatResult::WINGAME
      else
        combat_result=CombatResult::WIN
      end

    else
      applyBadConsequence(m)
      combat_result=CombatResult::LOSE
    end
    return combat_result
  end
  
  def makeTreasureVisible(t)
    can_i=canMakeTreasureVisible(t) 

    if(can_i)
      @visibleTreasures << t 
      @hiddenTreasures.delete(t)
    end
  end
  
  def discardVisibleTreasure(t)
    @visibleTreasures.delete(t)
    if (@pendingBadConsequence != nil && !@pendingBadConsequence.isEmpty)
      @pendingBadConsequence.substractVisibleTreasure(t)
    end
    dieIfNoTreasure
  end
  
  def discardHiddenTreasure(t)
    @hiddenTreasures.delete(t)
    if(@pendingBadConsequence != nil && !@pendingBadConsequence.isEmpty)
      @pendingBadConsequence.substractHiddenTreasure(t)
    end
    dieIfNoTreasures
  end
  
  def validState
     return ( @pendingBadConsequence.isEmpty && (@hiddenTreasures.size <= 4) )
  end
  
  def initTreasures
    return (@pendingBadConsequence == nil || @pendingBadConsequence.isEmpty) && @hiddenTreasures.length<=4
  end
  
  
  def initTreasures
    dealer=CardDealer.instance
    dice=Dice.instance

    bringToLife

    treasure=dealer.nextTreasure
    @hiddenTreasures << treasure
    number=dice.nextNumber

    if(number > 1)
      treasure=dealer.nextTreasure
      @hiddenTreasures << treasure
    end

    if(number == 6)
      treasure=dealer.nextTreasure
      @hiddenTreasures << treasure
    end
  end
  
  def stealTreasure

    if(@canISteal)
      can_you=@enemy.canYouGiveMeATreasure

      if(can_you)
        treasure=@enemy.giveMeATreasure
        @hiddenTreasures << treasure
        haveStolen
        return treasure

      else
        return nil
      end
    else
      return nil
    end
  end
  
  
  def giveMeATreasure
     @hiddenTreasures.at(rand(@hiddenTreasures.length + 1))
  end
  
  
  def canYouGiveMeATreasure
     !@hiddenTreasures.empty?
  end
  
  def haveStolen
     @canISteal = false
  end
 
  def discardAllTreasures
    v=@visibleTreasures

    v.each do |vi|
      discardVisibleTreasure(vi)
    end

    h= @hiddenTreasures

    h.each do |hi| 
      discardHiddenTreasure(hi) 
    end
  end
def setEnemy(enem)
    @enemy=enem
end
end

File: 2º/PDOO/Practias/Napakalaki/src/napakalaki/Player.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package napakalaki;

import java.util.ArrayList;
import java.util.Random;

/**
 *
 * @author arturo
 */
public class Player {
    static final int MAXLEVEL=10;
    
    private String name;
    private int level;
    private boolean dead = true;
    private boolean canISteal =true;
    private Player enemy;
    private ArrayList<Treasure> hiddenTreasures= new ArrayList<Treasure>();;
    private ArrayList<Treasure> visibleTreasures= new ArrayList<Treasure>();;
    private BadConsequence pendingBadConsequence;

    public Player(String name) {
        this.name = name;
    }

    public String getName() {
        return name;
    }
    
    private void bringTolife(){
        dead = false;
    }
    
    private int getCombatLevel(){
    
        int combat_level;
        combat_level=level;
        for(Treasure t:visibleTreasures){
            combat_level=combat_level+t.getBonus();
        }
        return combat_level;
        
    }
    
    private void incrementLevels(int l ){
        if( l > 0 ){
            level += l;
        }
    }
    
    private void decrementLevels(int l){
        if( l > 0 ){
            level -= l;
        }
        
    }
    
    private void setPendingBadConsequence(BadConsequence b){
        pendingBadConsequence = b;
        
    }
    
    private void applyPrize(Monster m){
        int nLevels = m.getLevelsGained();
        int nTreasures = m.getTreasuresGained();
        incrementLevels(level);
        
        if(nTreasures > 0){
            CardDealer dealer = CardDealer.getInstance();
            Treasure treasure;
            for(int i = 0; i < nTreasures; i++){
                treasure = dealer.nextTreasure();
                hiddenTreasures.add(treasure);
            }
        }
    }
    
    private void applyBadConsequence(Monster m){
        BadConsequence badConsequence = m.getBc();
        int nLevels = badConsequence.getLevels();
        decrementLevels(nLevels);
        BadConsequence pendingBad = badConsequence.adjustToFitTreasureLists(visibleTreasures, hiddenTreasures);
        setPendingBadConsequence(pendingBad);
    }
    
    private boolean canMakeTreasureVisible(Treasure T){          
        boolean ret = true;
        TreasureKind type = T.getType();
        
        if (type == TreasureKind.ONEHAND){
            if ((howManyVisibleTreasures(TreasureKind.BOTHHANDS) > 0) || (howManyVisibleTreasures(TreasureKind.ONEHAND) > 1))
                ret = false;
        }        
        else if (type == TreasureKind.BOTHHANDS){
            if ((howManyVisibleTreasures(TreasureKind.BOTHHANDS) > 0) || (howManyVisibleTreasures(TreasureKind.ONEHAND) > 0))
                ret = false;
        }
        else{
            if (howManyVisibleTreasures(type) > 0)
                ret = false;
        }
        return ret;
    }
    
    private int howManyVisibleTreasures(TreasureKind tKind){
        int i = 0;
        
        for (Treasure t : visibleTreasures){
            if( t.getType() == tKind )
                i++;
        }
        return i;
    }
    
    private void dieIfNoTreasures(){
        if( hiddenTreasures.isEmpty() && visibleTreasures.isEmpty() ){
            dead = true;
        }   
    }
    
    public boolean isDead(){
        return dead;
    }
    
    public ArrayList<Treasure> getHiddenTreasures(){
        return hiddenTreasures;
    }
    
    public ArrayList<Treasure> getVisibleTreasures(){
        return visibleTreasures;
    }
    public CombatResult combat(Monster m){
        int myLevel = getCombatLevel();
        int m_Level = m.getCombatLevel();
        CombatResult combatResult;
        
        if(!canISteal){
            Dice dice = Dice.getInstance();
            int n = dice.nextNumber();
            
            if(n<3){
                int e_Level = enemy.getCombatLevel();
                m_Level += e_Level;
            }
        }
        
        if(myLevel > m_Level){
            applyPrize(m);
            if(level >= MAXLEVEL)
                combatResult = CombatResult.WINGAME;
            else
                combatResult = CombatResult.WIN;           
        }
        else{
            applyBadConsequence(m);
            combatResult = CombatResult.LOSE;
        }
        
        return combatResult;
    }
    
    public void makeTreasureVisible(Treasure t){}
            
    public void discardVisibleTreasure(Treasure t){}
            
    public void  discardHiddenTreasure( Treasure t){}
    
    public boolean validState(){
        return pendingBadConsequence.isEmpty() && ( hiddenTreasures.size() <= 4 );
    }
            
    public void initTreasures(){
        CardDealer dealer = CardDealer.getInstance();
        Dice dice = Dice.getInstance();
        
        bringTolife();
        Treasure treasure = dealer.nextTreasure();
        hiddenTreasures.add(treasure);
        int number = dice.nextNumber();
        
        if(number > 1 && number < 6){
            treasure = dealer.nextTreasure();
            hiddenTreasures.add(treasure);           
        }
        if(number == 6){
            treasure = dealer.nextTreasure();
            hiddenTreasures.add(treasure);
        }
    
    }

    public int getLevels() {
        return level;
    }
            
    public Treasure stealTreasure(){
        Treasure t=null;
        if(canISteal()){
            if(enemy.canYouGiveMeATreasure()){
               t = enemy.giveMeATreasure();
               hiddenTreasures.add(t);
               haveStolen();
            }
        }
        return t;
    }     
    
    public void setEnemy(Player enemy){
        this.enemy = enemy;
    }
    
    private Treasure giveMeATreasure(){
            Random ran=new Random();
            return hiddenTreasures.get(ran.nextInt(hiddenTreasures.size()));
    }
    
    public boolean canISteal(){
        return canISteal;
    }
         
    private boolean canYouGiveMeATreasure(){
         return !hiddenTreasures.isEmpty();
    }    

    private void haveStolen(){
         canISteal = false;
    }
    
    public void discardAllTreasures(){
       Treasure treasure;
       for(int i = 0; i < visibleTreasures.size(); i++){
           treasure = visibleTreasures.get(i);
           discardVisibleTreasure(treasure);
       }
       for(int i = 0; i < hiddenTreasures.size(); i++){
           treasure = hiddenTreasures.get(i);
           discardHiddenTreasure(treasure);
        }
    
    }
}
